Use gdbm for ODBM_File/NDBM_File.  The compatability header is named
gdbm-ndbm.h on debian machines.  Explicitly link against -lgdbm_compat.

diff -Naur --exclude=debian perl-5.10.0.orig/Configure perl-5.10.0/Configure
--- perl-5.10.0.orig/Configure	2007-12-18 21:47:07.000000000 +1100
+++ perl-5.10.0/Configure	2007-12-21 22:09:05.000000000 +1100
@@ -20560,7 +20560,7 @@
 eval $inhdr
 
 : see if ndbm.h is available
-set ndbm.h t_ndbm
+set gdbm-ndbm.h t_ndbm
 eval $inhdr
 
 case "$t_ndbm" in
diff -Naur --exclude=debian perl-5.10.0.orig/ext/NDBM_File/NDBM_File.xs perl-5.10.0/ext/NDBM_File/NDBM_File.xs
--- perl-5.10.0.orig/ext/NDBM_File/NDBM_File.xs	2007-12-18 21:47:07.000000000 +1100
+++ perl-5.10.0/ext/NDBM_File/NDBM_File.xs	2007-12-21 22:09:05.000000000 +1100
@@ -1,7 +1,7 @@
 #include "EXTERN.h"
 #include "perl.h"
 #include "XSUB.h"
-#include <ndbm.h>
+#include <gdbm-ndbm.h>
 
 typedef struct {
 	DBM * 	dbp ;
diff -Naur --exclude=debian perl-5.10.0.orig/ext/NDBM_File/hints/linux.pl perl-5.10.0/ext/NDBM_File/hints/linux.pl
--- perl-5.10.0.orig/ext/NDBM_File/hints/linux.pl	2007-12-18 21:47:07.000000000 +1100
+++ perl-5.10.0/ext/NDBM_File/hints/linux.pl	2007-12-21 22:09:05.000000000 +1100
@@ -2,7 +2,4 @@
 # Prefer gdbm to avoid the broken ndbm in some distributions
 # (no null key support)
 # Jonathan Stowe <gellyfish@gellyfish.com>
-use Config;
-use ExtUtils::Liblist;
-($self->{LIBS}) = ExtUtils::Liblist->ext('-lgdbm -lgdbm_compat')
-	if $Config{libs} =~ /(?:^|\s)-lgdbm(?:\s|$)/;
+$self->{LIBS} = ['-lgdbm_compat'];
diff -Naur --exclude=debian perl-5.10.0.orig/ext/ODBM_File/hints/linux.pl perl-5.10.0/ext/ODBM_File/hints/linux.pl
--- perl-5.10.0.orig/ext/ODBM_File/hints/linux.pl	2007-12-18 21:47:07.000000000 +1100
+++ perl-5.10.0/ext/ODBM_File/hints/linux.pl	2007-12-21 22:10:25.000000000 +1100
@@ -1,8 +1,2 @@
 # uses GDBM dbm compatibility feature - at least on SuSE 8.0
-$self->{LIBS} = ['-lgdbm'];
-
-# Debian/Ubuntu have /usr/lib/libgdbm_compat.so.3* but not this file,
-# so linking may fail
-if (-e '/usr/lib/libgdbm_compat.so' or -e '/usr/lib64/libgdbm_compat.so') {
-    $self->{LIBS}->[0] .= ' -lgdbm_compat';
-}
+$self->{LIBS} = ['-lgdbm_compat'];
